tessen

```diff
@@ -334,20 +335,17 @@ is_installed() {
 }
 
 set_bemenu() {
-  local -a bmn_opt=("-i -l 10 -w --scrollbar=autohide -n")
+  local -a bmn_opt=("-i" "-l" "10" "-w" "--scrollbar=autohide" "-n")
   readonly tsn_backend="bemenu"
-  # use BEMENU_OPTS if set by the user, otherwise use $bmn_opt
-  export BEMENU_OPTS="${BEMENU_OPTS:-${bmn_opt[*]}}"
```

Comment:
`bemenu` doesn't use any method of configuration besides command line flags and environment variables. `tessen` is already using the 1st method and by removing the 2nd method, environment variables, it becomes impossible for a user to customize the behavior of `bemenu`. What if a user wants to see 15 lines instead of 10?

I don't think `BEMENU_OPTS` should be removed.

Reply:
`BEMENU_OPTS` is global and it is used by `bemenu` internally. For example I only keep the defaults (color theme, font, height) in `BEMENU_OPTS` and use case specific settings are passed as arguments, when I want to run application one line menu is more convinient which for password searching do not work best.
If you want to allow user to configure `bemenu`, I would opt for environment variable, eg: `TSN_BEMENU_OPTS` or I would pass the configuration in backend switch:
```
tessen --backend='bemenu -l10 ...'
```
and remove all configuration code for menu backends, something similar to solution from [networkmanager-dmenu](https://github.com/firecat53/networkmanager-dmenu)

Reply:
> I only keep the defaults (color theme, font, height) in `BEMENU_OPTS` and use case specific settings are passed as arguments, when I want to run application one line menu is more convenient which for password searching do not work best.

Hmm, makes sense.

> If you want to allow user to configure `bemenu`, I would opt for environment variable, eg: `TSN_BEMENU_OPTS` or I would pass the configuration in backend switch

I don't want to complicate things even more by adding another environment variable for a specific backend. Removing the backend configuration sounds good and this is something I want to do anyways as mentioned in #4. At this moment, support for `fuzzel` is missing from `tessen` because of this issue.

I went for the whitelist approach because I didn't want the possibility of invalid input from users which would lead to `tessen` not working as expected. I also didn't want users to be concerned with configuring the backend program they were using but I don't think expecting users to do that is unreasonable given that they're using a tool like `tessen`.

Would you be interested in raising a PR to fix #4? 🙂 This would be a breaking change so it would need a new major release. It'll need some significant changes to the code and the documentation. Let me know what you think.

Reply:
Yes, I can do this. It can be done without breaking backward compatibility, so only minor release.
I can do the coding. Docs are on your side ;)

Reply:
Awesome, sounds good. Please raise a new PR if you decide to work on #4.

Reply:
By the way, would you like to explain how you intend to preserve backward compatibility? Right now, if a `tessen` user installs a backend, he doesn't need to configure it and it would work out of the box because `tessen` provides backend options for all the backends that it supports.

We can either make the `--backend` flag take in the name of the backend and its options like you've mentioned before or we could introduce another command line option to take in the backend options, something like `--backend-opts`. Either way, we would break backward compatibility.

Reply:
When you use `--backend` option or `TESSEN_BACKEND` variable script will parse them, if not it will try to use one of previously supported backends with defaults, also if user pass only the backend name, script will add existing defaults.

Reply:
Hmm, I was thinking about breaking backward compatibility to fix #4 and just removing any default options and leaving everything in the hands of the user.

I guess the best method would be to parse the `--backend` option like you said and check if the user has specified any options. If yes, `tessen` won't specify any backend options of its own. If only a backend name is mentioned and no options have been provided, `tessen` will try to provide a default option if it recognizes the backend.

So if a user writes `tessen -b fuzzel`, `tessen` should try and provide a default option and use `fuzzel -d`. If a user writes `tessen -b "fuzzel -d"`, `tessen` should use the backend option as is and not check anything.

Sounds good?

Reply:
> So if a user writes `tessen -b fuzzel`, `tessen` should try and provide a default option and use `fuzzel -d`. If a user writes `tessen -b "fuzzel -d"`, `tessen` should use the backend option as is and not check anything.

This is exactly what I had in mind.

Reply:
Nice, I'll wait for your PR. 🙂

Reply:
Done

Reply:
Wow, that was fast. 😮 

Let me go through the changes.